packages/controller/lib/zipFiles.js

```diff
@@ -168,80 +168,64 @@ function _checkDir(objects, id, root, parts, options, callback) {
-        const parts = fName.split('/');
-        parts.pop();
-        _checkDir(objects, id, '', parts, options, () =>
-            objects.writeFile(id, name + filename, data, options, err => callback(_err || err))
-        );
-    } catch (e) {
-        callback(e.message);
     }
+    const fName = name + filename;
+    const parts = fName.split('/');
+    parts.pop();
+    return new Promise((resolve, reject) =>
+        _checkDir(objects, id, '', parts, options, () =>
+            objects.writeFile(id, name + filename, data, options, err =>
+                _err || err ? reject(_err || err) : resolve()
+            )
+        )
+    );
 }
 
-async function writeDirAsZip(objects, id, name, data, options, callback) {
+async function writeDirAsZip(objects, id, name, data, options) {
     JSZip = JSZip || require('jszip');
     const zip = new JSZip();
 
     options = options || {};
 
     let adapter = id;
-    if (adapter.indexOf('.') !== -1) {
+    if (adapter.includes('.')) {
         adapter = id.split('.')[0];
     }
 
     // try to load processor of adapter
     try {
-        options.parse = require(tools.appName + '.' + adapter + '/lib/convert.js').parse;
+        options.parse = require(`${tools.appName}.${adapter}/lib/convert.js`).parse;
     } catch {
         // OK
     }
 
-    try {
-        await zip.loadAsync(data);
-        let count = 0;
-        const error = [];
-        if (name[name.length - 1] !== '/') {
-            name += '/';
-        }
-        for (const filename of Object.keys(zip.files)) {
-            if (!filename || filename[filename.length - 1] === '/') {
-                continue;
-            }
-            count++;
-            try {
-                _writeOneFile(objects, zip, id, name, filename, options, err => {
-                    err && error.push('Cannot write file "' + filename + '":' + err.toString());
-
-                    if (!--count && callback) {
-                        callback(error.length ? error.join(', ') : null);
-                        callback = null;
-                    }
-                });
-            } catch (error) {
-                if (callback) {
-                    callback(error.toString());
-                    callback = null;
-                }
-            }
+    await zip.loadAsync(data);
+    const errors = [];
+    if (name[name.length - 1] !== '/') {
+        name += '/';
+    }
+    for (const filename of Object.keys(zip.files)) {
+        if (!filename || filename[filename.length - 1] === '/') {
+            continue;
         }
-    } catch (error) {
-        if (callback) {
-            callback(error.toString());
-            callback = null;
+        try {
+            await _writeOneFile(objects, zip, id, name, filename, options);
+        } catch (error) {
+            errors.push(`Cannot write file "${filename}": ${error.toString()}`);
         }
     }
+    if (errors.length) {
+        throw errors.join(', ');
```

Comment:
Let us throw a real error instead of a string here. Normally eslint should complain here.

Reply:
Done